File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        res = []
        digitToChar = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "qprs",
            "8": "tuv",
            "9": "wxyz",
        }
        def backtrack(i, curr):
            if len(curr) == len(digits):
                res.append(curr)
                return
            for char in digitToChar[digits[i]]:
                curr = curr + char
                backtrack(i+1, curr)
                curr = curr[:-1]
        if digits:
            backtrack(0, "")
        return res

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string> result;
    unordered_map<char, string> hm;

    vector<string> letterCombinations(string digits) {
        if(digits.empty()){
            return {};
        }
        hm['2'] = "abc";
        hm['3'] = "def";
        hm['4'] = "ghi";
        hm['5'] = "jkl";
        hm['6'] = "mno";
        hm['7'] = "pqrs";
        hm['8'] = "tuv";
        hm['9'] = "wxyz";
        string temp;
        dfs(temp, 0, digits);
        return result;
        
    }

    void dfs(string& temp, int i,  string digits) {
        if (i == digits.size()) {
            result.push_back(temp);
            return;
        }
    
        for(auto val: hm[digits[i]]) {
            temp.push_back(val);
            dfs(temp, i + 1,   digits);
            temp.pop_back();
        }
    }
};